### PR TITLE
Prevent loading images that would require too much memory.

### PR DESCRIPTION
**Pull request by fancycode**
For most image formats, the header specifies the width/height. PHP allocates an image object from that size, even if the actual
image data is much smaller. This image object size is not limited by the limit configured in PHP.

The memory limit can be configured through config setting "preview_max_memory" and defaults to 128 MBytes which should be enough for most images without filling up all memory.

Signed-off-by: Joachim Bauch <bauch@struktur.de>

### CODE REVIEW

File: lib/private/legacy/OC_Image.php [hunk 1]
```diff
@@ -563,6 +567,71 @@ public function loadFromFileHandle($handle) {
...
 
+	/**
+	 * Check if allocating an image with the given size is allowed.
+	 *
+	 * @param int $width The image width.
+	 * @param int $height The image height.
+	 * @return bool true if allocating is allowed, false otherwise
+	 */
+	private function checkImageMemory($width, $height) {
+		$memory_limit = (int) $this->config->getAppValue('preview', 'memory_limit', (string) self::DEFAULT_MEMORY_LIMIT);
```
Comment (nickvergessen) [inline lib/private/legacy/OC_Image.php:578]: From my POV this should go into the config file:
https://github.com/nextcloud/server/blob/4b36f9d92f886609e34f5a180f38e6034aacbd79/config/config.sample.php#L1001-L1044

Comment (nickvergessen) [reply lib/private/legacy/OC_Image.php:578]:
```suggestion
		$memory_limit = (int) $this->config->getSystemValueInt('preview_max_memory', self::DEFAULT_MEMORY_LIMIT);
```

**Review by ChristophWurst — APPROVED** [review]

File: lib/private/legacy/OC_Image.php [hunk 2]
```diff
@@ -563,6 +567,71 @@ public function loadFromFileHandle($handle) {
...
+		if ($memory_limit <= 0) {
+			// Not limited.
+			return true;
+		}
+
+		// Assume 32 bits per pixel.
+		if ($width * $height * 4 > $memory_limit) {
+			$this->logger->warning('Image size of ' . $width . 'x' . $height . ' would exceed allowed memory limit of ' . $memory_limit);
```
Comment (blizzz) [inline lib/private/legacy/OC_Image.php:586]: The level seems too high, info or debug would be sufficient (especially when you cannot do too much about it).

File: lib/private/legacy/OC_Image.php (hunk 1 above)
Comment (nickvergessen) [reply lib/private/legacy/OC_Image.php:578]: So I was going to do this and found this on the way:
https://github.com/nextcloud/server/blob/168c673755eca62008d9bfa81d0185beba1ff24b/config/config.sample.php#L1037-L1044

Should be exactly what you are trying to do? Or is your file size small and the image dimension still bigger?

**Comment by skjnldsv** [comment]
Hey @fancycode , could you address the comments ? :)
Happy new year! :tada:

**Comment by fancycode** [comment]
> Hey @fancycode , could you address the comments ? :) Happy new year! tada

Happy new year to you, too. I just came back from vacation today and are going through my inbox. Will address the comments as soon as possible.

File: lib/private/legacy/OC_Image.php (hunk 1 above)
Comment (fancycode) [reply lib/private/legacy/OC_Image.php:578]: This is something else. See the image added to the test which is 103 bytes but will cause PHP to allocate multiple gigabytes of memory.

File: lib/private/legacy/OC_Image.php (hunk 2 above)
Comment (fancycode) [reply lib/private/legacy/OC_Image.php]: Changed to debug.

**Comment by fancycode** [comment]
All comments addressed. Commits should be squashed and the text of the first commit should be adjusted (limit is now read from config.php instead of the app config) - I can do this before merging.

**Review by nickvergessen — APPROVED** [review]

**Comment by fancycode** [comment]
Commits are squashed and the message is updated, so this is ready to merge from my side.

**Review by blizzz — APPROVED** [review]

**Comment by fancycode** [comment]
Thanks for merging. Probably would be good to also backport this change to all versions that still receive security updates.

**Pull request merged**